flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/topic/range/FixedKeysRangeGenerator.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Override
+    public List<TopicRange> range(TopicMetadata metadata, int parallelism) {
+        return ranges;
+    }
+
+    @Override
+    public KeySharedMode keyShareMode(TopicMetadata metadata, int parallelism) {
+        return sharedMode;
+    }
+
+    public static FixedKeysRangeGeneratorBuilder builder() {
+        return new FixedKeysRangeGeneratorBuilder();
+    }
+
+    /** The builder for {@link FixedKeysRangeGenerator}. */
+    @PublicEvolving
+    public static class FixedKeysRangeGeneratorBuilder {
+
+        private final SortedSet<Integer> keyHashes = new TreeSet<>();
+        private KeySharedMode sharedMode = KeySharedMode.JOIN;
+
+        private FixedKeysRangeGeneratorBuilder() {
+            // No public for builder
+        }
+
+        /**
+         * Some {@link Message} in Pulsar may not have {@link Message#getOrderingKey()} or {@link
+         * Message#getKey()}, use this method for supporting consuming such messages.
+         */
+        public FixedKeysRangeGeneratorBuilder supportNullKey() {
+            keyHashes.add(keyHash(NONE_KEY));
+            return this;
+        }
+
+        /**
+         * If you set the message key by using {@link PulsarMessageBuilder#key(String)} or {@link
+         * TypedMessageBuilder#key(String)}, use this method for supporting consuming such messages.
+         */
+        public FixedKeysRangeGeneratorBuilder key(String key) {
+            keyHashes.add(keyHash(key));
+            return this;
+        }
+
+        /** Same as the {@link #key(String)}, support setting multiple keys in the same time. */
+        public FixedKeysRangeGeneratorBuilder keys(Collection<String> someKeys) {
+            checkNotNull(someKeys);
+            for (String someKey : someKeys) {
+                keyHashes.add(keyHash(someKey));
+            }
+            return this;
+        }
+
+        /**
+         * If you set the message key by using {@link TypedMessageBuilder#keyBytes(byte[])}, use
+         * this method for supporting consuming such messages.
+         */
+        public FixedKeysRangeGeneratorBuilder keyBytes(byte[] keyBytes) {
+            keyHashes.add(keyBytesHash(keyBytes));
+            return this;
+        }
+
+        /**
+         * Pulsar's ordering key is prior to the message key. If you set the ordering key by using
+         * {@link PulsarMessageBuilder#orderingKey(byte[])} or {@link
+         * TypedMessageBuilder#orderingKey(byte[])}, use this method for supporting consuming such
+         * messages.
+         */
+        public FixedKeysRangeGeneratorBuilder orderingKey(byte[] keyBytes) {
+            keyHashes.add(keyHash(keyBytes));
+            return this;
+        }
+
+        /** Override the default {@link KeySharedMode#JOIN} to the mode your have provided. */
+        public FixedKeysRangeGeneratorBuilder keySharedMode(KeySharedMode sharedMode) {
+            this.sharedMode = sharedMode;
+            return this;
+        }
```

Comment:
For what reason we can use a different mode here? I think for consuming fixed key range we should always use `JOIN`...unless the fixed range is full range? Then users should use `FullRangeGenerator` or `SplitRangeGenerator` instead.

Reply:
If we have 6 kind of keys in production. 3 keys are used in one flink application and others are used in another flink application. Use `SPLIT` mode can consume the messages in different readers. I don't want to choose for the end user but to give them the choice to use the key shared mode.